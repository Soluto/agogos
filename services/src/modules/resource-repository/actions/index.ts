import applyResourceGroupUpdates from './update';
import applyResourceGroupDeletions from './delete';

export { applyResourceGroupUpdates, applyResourceGroupDeletions };
