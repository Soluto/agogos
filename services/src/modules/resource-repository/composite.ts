import {ResourceRepository, FetchLatestResult, ResourceGroup} from './types';
import {applyResourceGroupUpdates} from './updates';

export class CompositeResourceRepository implements ResourceRepository {
    constructor(protected repositories: ResourceRepository[]) {}

    async fetchLatest(): Promise<FetchLatestResult> {
        const results = await Promise.all(this.repositories.map(repo => repo.fetchLatest()));

        return results.reduce((res1, res2) => ({
            isNew: res1.isNew || res2.isNew,
            resourceGroup: applyResourceGroupUpdates(res1.resourceGroup, res2.resourceGroup),
        }));
    }

    getResourceGroup(): ResourceGroup {
        const rgs = this.repositories.map(r => r.getResourceGroup());

        return rgs.reduce((rg1, rg2) => applyResourceGroupUpdates(rg1, rg2));
    }

    async update(): Promise<void> {
        throw new Error('Multiplexed resource repository cannot handle updates');
    }

    async writePolicyAttachment(): Promise<void> {
        throw new Error('Multiplexed resource repository cannot handle updates');
    }

    public getPolicyAttachment(filename: string): Buffer {
        for (const repo of this.repositories) {
            const policyAttachment = repo.getPolicyAttachment(filename);
            if (policyAttachment) return policyAttachment;
        }

        throw new Error(`Policy attachment with the filename ${filename} was not found`);
    }

    public async initializePolicyAttachments() {
        await Promise.all(this.repositories.map(repo => repo.initializePolicyAttachments()));
    }
}
