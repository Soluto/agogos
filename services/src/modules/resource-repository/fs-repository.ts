import * as envVar from 'env-var';
import { FileSystemStorage } from '../storage';
import { ResourceRepository } from '.';

export class FileSystemResourceRepository extends ResourceRepository {
  protected policyAttachmentsFolderInitialized = false;

  constructor(
    protected storage: FileSystemStorage,
    protected resourceFilePath: string,
    protected policyAttachmentsFolderPath: string,
    protected isRegistry = false
  ) {
    super(storage, resourceFilePath, policyAttachmentsFolderPath, isRegistry);
  }

  async writePolicyAttachment(filename: string, content: Buffer): Promise<void> {
    await this.initializePolicyAttachmentsFolder();
    await super.writePolicyAttachment(filename, content);
  }

  async listPolicyAttachments() {
    await this.initializePolicyAttachmentsFolder();
    return super.listPolicyAttachments();
  }

  private async initializePolicyAttachmentsFolder() {
    if (this.policyAttachmentsFolderInitialized) return;

    await this.storage.mkdir(this.policyAttachmentsFolderPath, { recursive: true });
    this.policyAttachmentsFolderInitialized = true;
  }

  static fromEnvironment(options: { isRegistry: boolean } = { isRegistry: false }) {
    const resourceFilePath = envVar.get('FS_RESOURCE_REPOSITORY_PATH').required().asString();
    const policyAttachmentsFolderPath = envVar
      .get('FS_REPOSITORY_POLICY_ATTACHMENTS_FOLDER_PATH')
      .required()
      .asString();

    const fsStorage = new FileSystemStorage();
    return new FileSystemResourceRepository(
      fsStorage,
      resourceFilePath,
      policyAttachmentsFolderPath,
      options.isRegistry
    );
  }
}
