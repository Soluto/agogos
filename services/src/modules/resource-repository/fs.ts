import {ResourceGroup, ResourceRepository} from '.';
import * as envVar from 'env-var';
import {promises as fs} from 'fs';
import * as path from 'path';
import pLimit from 'p-limit';
import * as config from '../config';
import logger from '../logger';
import {FetchLatestResult} from './types';

export class FileSystemResourceRepository implements ResourceRepository {
    protected current?: {mtime: number; rg: ResourceGroup};
    protected policyAttachmentsDirInitialized = false;
    protected policyAttachments: {[filename: string]: Buffer} = {};
    protected policyAttachmentsRefreshedAt?: Date;

    constructor(protected pathToFile: string, protected policyAttachmentsFolderPath: string) {}

    async fetchLatest(): Promise<FetchLatestResult> {
        const stats = await fs.stat(this.pathToFile);

        if (this.current && stats.mtimeMs === this.current.mtime) {
            return {isNew: false, resourceGroup: this.current.rg};
        }

        const contents = await fs.readFile(this.pathToFile, 'utf8');
        const rg = JSON.parse(contents) as ResourceGroup;
        this.current = {mtime: stats.mtimeMs, rg};

        return {isNew: true, resourceGroup: rg};
    }

    async update(rg: ResourceGroup): Promise<void> {
        await fs.writeFile(this.pathToFile, JSON.stringify(rg));
    }

    async writePolicyAttachment(filename: string, content: Buffer): Promise<void> {
        await this.initializePolicyAttachmentsDir();

        const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
        await fs.writeFile(filePath, content);
    }

    public getPolicyAttachment(filename: string): Buffer {
        return this.policyAttachments[filename];
    }

    public async initializePolicyAttachments() {
        try {
            await this.refreshPolicyAttachments();
        } catch (err) {
            logger.fatal({err}, 'Failed fetching fs policy attachments on startup');
            throw err;
        }

        setInterval(async () => {
            try {
                await this.refreshPolicyAttachments();
            } catch (err) {
                logger.error(
                    {err},
                    `Failed refreshing fs policy attachments, last successful refresh was at ${this.policyAttachmentsRefreshedAt}`
                );
            }
        }, config.resourceUpdateInterval);
    }

    private async refreshPolicyAttachments() {
        const newRefreshedAt = new Date();

        const allAttachments = await this.getPolicyAttachmentsList();
        const attachmentsToRefresh = allAttachments
            .filter(a => this.shouldRefreshPolicyAttachment(a))
            .map(a => a.filename);

        if (attachmentsToRefresh.length > 0) {
            const newAttachments = await this.getPolicyAttachments(attachmentsToRefresh);
            newAttachments.forEach(a => (this.policyAttachments[a.filename] = a.content));
        }

        this.policyAttachmentsRefreshedAt = newRefreshedAt;
    }

    private shouldRefreshPolicyAttachment({filename, updatedAt}: {filename: string; updatedAt: Date}) {
        if (!this.policyAttachments[filename]) return true;
        if (!this.policyAttachmentsRefreshedAt) return true;

        return updatedAt > this.policyAttachmentsRefreshedAt;
    }

    private async getPolicyAttachmentsList(): Promise<{filename: string; updatedAt: Date}[]> {
        const filenames = await fs.readdir(this.policyAttachmentsFolderPath);
        const limit = pLimit(10);

        return Promise.all(
            filenames.map(filename => {
                return limit(async () => {
                    const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
                    const stats = await fs.stat(filePath);

                    const updatedAt = stats.mtime;
                    return {filename, updatedAt};
                });
            })
        );
    }

    private async getPolicyAttachments(filenames: string[]): Promise<{filename: string; content: Buffer}[]> {
        const limit = pLimit(10);

        return Promise.all(
            filenames.map(filename => {
                return limit(async () => {
                    const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
                    const content = await fs.readFile(filePath);

                    return {filename, content};
                });
            })
        );
    }

    private async initializePolicyAttachmentsDir() {
        if (this.policyAttachmentsDirInitialized) return;

        await fs.mkdir(this.policyAttachmentsFolderPath, {recursive: true});
        this.policyAttachmentsDirInitialized = true;
    }

    static fromEnvironment() {
        const pathToFile = envVar
            .get('FS_RESOURCE_REPOSITORY_PATH')
            .required()
            .asString();
        const policyAttachmentsFolderPath = envVar
            .get('FS_REPOSITORY_POLICY_ATTACHMENTS_FOLDER_PATH')
            .required()
            .asString();

        return new FileSystemResourceRepository(pathToFile, policyAttachmentsFolderPath);
    }
}
