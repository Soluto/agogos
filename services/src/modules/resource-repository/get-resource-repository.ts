import * as config from '../config';
import logger from '../logger';
import {
  CompositeResourceRepository,
  FileSystemResourceRepository,
  IResourceRepository,
  ResourceRepository,
  S3ResourceRepository,
} from '.';

export default function getResourceRepository(options?: { isRegistry: boolean }): IResourceRepository {
  const repositories: ResourceRepository[] = [];

  if (config.useS3ResourceRepository) {
    repositories.push(S3ResourceRepository.fromEnvironment(options));
  }

  if (config.useFileSystemResourceRepository) {
    repositories.push(FileSystemResourceRepository.fromEnvironment(options));
  }

  switch (repositories.length) {
    case 0:
      logger.fatal('Must enable at least one resource repository');
      throw new Error('Must enable at least one resource repository');
    case 1:
      return repositories[0];
    default:
      if (options?.isRegistry) {
        logger.fatal('Registry cannot have more than one resource repository');
        throw new Error('Registry cannot have more than one resource repository');
      }
      return new CompositeResourceRepository(repositories);
  }
}
