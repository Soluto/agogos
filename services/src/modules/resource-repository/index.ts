export * from './types';

export { ResourceRepository } from './resource-repository';
export { CompositeResourceRepository } from './composite';
export { S3ResourceRepository } from './s3-repository';
export { FileSystemResourceRepository } from './fs-repository';
export { pollForUpdates } from './stream';
export { applyResourceGroupUpdates, applyResourceGroupDeletions } from './actions';
export { default as getResourceRepository } from './get-resource-repository';
