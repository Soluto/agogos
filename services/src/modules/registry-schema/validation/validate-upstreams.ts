import { ApolloError } from 'apollo-server-core';
import { GraphQLError } from 'graphql';
import * as _ from 'lodash';
import { ResourceMetadata, Upstream } from '../../resource-repository';

export default function validateUpstreams(upstreams: Upstream[]) {
  const errors: GraphQLError[] = [];

  const existingHosts = new Map<string, Upstream>();
  for (const upstream of upstreams) {
    validateUpstreamTemplate(errors, upstream, upstreams, new Set<ResourceMetadata>());

    if (!upstream?.isTemplate && !upstream.host && !upstream.sourceHosts) {
      errors.push(
        new ApolloError(
          'Upstream should have either sourceHosts or host property or to be a template',
          'INVALID_UPSTREAM_CONFIGURATION',
          {
            metadata: upstream.metadata,
          }
        )
      );
      continue;
    }

    if (upstream.host && upstream.sourceHosts) {
      errors.push(
        new ApolloError('Upstream should have either sourceHosts or host property', 'INVALID_UPSTREAM_CONFIGURATION', {
          host: upstream.host,
          sourceHosts: upstream.sourceHosts,
          metadata: upstream.metadata,
        })
      );
      continue;
    }

    const validateHost = (host: string) => {
      const existingUpstream = existingHosts.get(host);

      if (typeof existingUpstream === 'undefined') existingHosts.set(host, upstream);
      else {
        errors.push(
          new ApolloError('Duplicate host found on upstream', 'DUPLICATE_UPSTREAM_FOUND', {
            host,
            upstreams: [upstream.metadata, existingUpstream.metadata],
          })
        );
      }
    };

    if (upstream.host) validateHost(upstream.host);
    if (upstream.sourceHosts) upstream.sourceHosts.forEach(validateHost);
  }

  const existingResourceAuthorityPairs = new Map<string, Upstream>();
  for (const upstream of upstreams) {
    if (!upstream.auth) continue;

    const upstreamKey = JSON.stringify(upstream.auth.activeDirectory);
    const existingUpstream = existingResourceAuthorityPairs.get(upstreamKey);
    if (typeof existingUpstream === 'undefined') {
      existingResourceAuthorityPairs.set(upstreamKey, upstream);
    } else {
      errors.push(
        new ApolloError('Duplicate authority+resource found on upstream', 'DUPLICATE_UPSTREAM_FOUND', {
          authority: upstream.auth.activeDirectory,
          upstreams: [upstream.metadata, existingUpstream.metadata],
        })
      );
    }
  }

  return errors;
}

function validateUpstreamTemplate(
  errors: GraphQLError[],
  upstream: Upstream,
  upstreams: Upstream[],
  descendants: Set<ResourceMetadata>
) {
  if (!upstream) return;
  if (!upstream.fromTemplate) return;
  if (descendants.has(upstream.metadata)) {
    errors.push(
      new ApolloError('Upstream circular reference', 'UPSTREAM_CIRCULAR_REFERENCE', {
        metadata: upstream.metadata,
      })
    );
    return;
  }

  descendants.add(upstream.metadata);

  const upstreamTemplate = upstreams.find(u => _.isEqual(u.metadata, upstream.fromTemplate));
  if (!upstreamTemplate) {
    errors.push(
      new ApolloError('Upstream has non-existent upstream template', 'INVALID_UPSTREAM_CONFIGURATION', {
        metadata: upstream.metadata,
        template: upstream.fromTemplate,
      })
    );
    return;
  }
  validateUpstreamTemplate(errors, upstreamTemplate, upstreams, descendants);
}
