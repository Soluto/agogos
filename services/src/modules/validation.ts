import { ApolloError } from 'apollo-server-fastify';
import { GraphQLError } from 'graphql';
import { ResourceGroup, Upstream, Resource } from './resource-repository';

// Valid graphql name with the addition of dashes (http://spec.graphql.org/June2018/#sec-Names)
const validNameRegex = /^[A-Z_a-z-][\w-]*$/;

const resourceTypesWithMetadata = new Set(['schemas', 'upstreams', 'upstreamClientCredentials', 'policies']);
type ResourceTypeWithMetadata = 'schemas' | 'upstreams' | 'upstreamClientCredentials' | 'policies';

function validateUpstreams(upstreams: Upstream[]) {
  const errors: GraphQLError[] = [];

  const existingHosts = new Map<string, Upstream>();
  for (const upstream of upstreams) {
    const existingUpstream = existingHosts.get(upstream.host);
    if (typeof existingUpstream === 'undefined') {
      existingHosts.set(upstream.host, upstream);
    } else {
      errors.push(
        new ApolloError('Duplicate host found on upstream', 'DUPLICATE_UPSTREAM_FOUND', {
          host: upstream.host,
          upstreams: [upstream.metadata, existingUpstream.metadata],
        })
      );
    }
  }

  const existingResourceAuthorityPairs = new Map<string, Upstream>();
  for (const upstream of upstreams) {
    if (!upstream.auth) continue;

    const upstreamKey = JSON.stringify(upstream.auth.activeDirectory);
    const existingUpstream = existingResourceAuthorityPairs.get(upstreamKey);
    if (typeof existingUpstream === 'undefined') {
      existingResourceAuthorityPairs.set(upstreamKey, upstream);
    } else {
      errors.push(
        new ApolloError('Duplicate authority+resource found on upstream', 'DUPLICATE_UPSTREAM_FOUND', {
          authority: upstream.auth.activeDirectory,
          upstreams: [upstream.metadata, existingUpstream.metadata],
        })
      );
    }
  }

  return errors;
}

function validateMetadataCharacters(rg: ResourceGroup) {
  const errors: GraphQLError[] = [];

  for (const [resourceType, resources] of Object.entries(rg) as [string, Resource[]][]) {
    if (!resourceTypesWithMetadata.has(resourceType)) continue;

    for (const resource of resources) {
      const { name, namespace } = resource.metadata;

      if (!validNameRegex.test(namespace)) {
        errors.push(
          new ApolloError(
            `Invalid characters found in namespace name ${namespace} of resource type ${resourceType}, allowed characters are /${validNameRegex.source}/`,
            'INVALID_CHARACTERS_IN_NAME',
            { resource: { type: resourceType, namespace, name } }
          )
        );
      }

      if (!validNameRegex.test(name)) {
        errors.push(
          new ApolloError(
            `Invalid characters found in resource name ${name} of resource type ${resourceType}, allowed characters are /${validNameRegex.source}/`,
            'INVALID_CHARACTERS_IN_NAME',
            { resource: { type: resourceType, namespace, name } }
          )
        );
      }
    }
  }

  return errors;
}

// Verifies namespaces and names to not have two similar ones differing only between
// underscores and dashes (e.g. 'some-ns' and 'some_ns' would return a conflict error)
function validateMetadataNameConflicts(rg: ResourceGroup) {
  const errors: GraphQLError[] = [];
  const existingNamespaces = new Map<string, string>();
  const existingNamesByNamespaceAndResource = {
    schemas: new Map<string, Map<string, string>>(),
    upstreams: new Map<string, Map<string, string>>(),
    upstreamClientCredentials: new Map<string, Map<string, string>>(),
    policies: new Map<string, Map<string, string>>(),
  };

  for (const [resourceType, resources] of Object.entries(rg) as [string, Resource[]][]) {
    if (!resourceTypesWithMetadata.has(resourceType)) continue;

    for (const resource of resources) {
      const { name, namespace } = resource.metadata;

      const namespaceWithUnderscores = namespace.replace(/-/g, '_');
      const existingNamespace = existingNamespaces.get(namespaceWithUnderscores);
      if (!existingNamespace) {
        existingNamespaces.set(namespaceWithUnderscores, namespace);
      } else {
        if (existingNamespace !== namespace) {
          errors.push(
            new ApolloError(
              `Namespace name conflict found between ${existingNamespace} and ${namespace}, they have to either match exactly or have a difference in non-special characters`,
              'NAME_CONFLICT',
              { resource: { type: resourceType, namespace, name } }
            )
          );
        }
      }

      const nameWithUnderscores = name.replace(/-/g, '_');
      const existingNamesByNamespace = existingNamesByNamespaceAndResource[resourceType as ResourceTypeWithMetadata];
      let existingNames = existingNamesByNamespace.get(namespace);
      if (!existingNames) {
        existingNames = new Map<string, string>();
        existingNamesByNamespace.set(namespace, existingNames);
      }

      const existingName = existingNames.get(nameWithUnderscores);
      if (!existingName) {
        existingNames.set(nameWithUnderscores, name);
      } else {
        if (existingName !== name) {
          errors.push(
            new ApolloError(
              `Resource name conflict found between ${existingName} and ${name} of resource type ${resourceType} in namespace ${namespace}, they have to either match exactly or have a difference in non-special characters`,
              'NAME_CONFLICT',
              { resource: { type: resourceType, namespace, name } }
            )
          );
        }
      }
    }
  }

  return errors;
}

export function validateResourceGroupOrThrow(rg: ResourceGroup) {
  const errors: GraphQLError[] = [];

  if (rg.upstreams.length > 0) {
    errors.push(...validateUpstreams(rg.upstreams));
  }

  errors.push(...validateMetadataCharacters(rg));
  errors.push(...validateMetadataNameConflicts(rg));

  if (errors.length > 0) {
    throw new ApolloError('Resource validation failed', 'RESOURCE_VALIDATION_FAILURE', {
      errors: errors.map(err => (err instanceof GraphQLError ? err : new GraphQLError(err!.message))),
    });
  }
}
