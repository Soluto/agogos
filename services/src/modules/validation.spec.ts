import GraphQLErrorSerializer from '../../tests/utils/graphql-error-serializer';
import { validateResourceGroupOrThrow } from './validation';
import { ResourceGroup, PolicyType } from './resource-repository';

beforeAll(() => {
  expect.addSnapshotSerializer(GraphQLErrorSerializer);
});

describe('validateMetadataCharacters - Validation that namespaces and names contain only valid gql chars and dashes', () => {
  let rg: ResourceGroup;

  beforeEach(() => {
    rg = {
      schemas: [{ metadata: { namespace: 'some-ns', name: 'special_not_allowed_here' }, schema: 'schema' }],
      upstreams: [],
      upstreamClientCredentials: [],
      policies: [{ metadata: { namespace: 'other-ns', name: 'valid-chars_only' }, type: PolicyType.opa, code: 'code' }],
      remoteSchemas: [],
    };
  });

  it("doesn't throw when names are valid", () => {
    expect(() => validateResourceGroupOrThrow(rg)).not.toThrow();
  });

  it('throws when namespace names contain invalid chars', () => {
    rg.schemas[0].metadata.namespace = 'special_not_allowed_@@';
    expect.assertions(1);

    try {
      validateResourceGroupOrThrow(rg);
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });

  it('throws when resource names contain invalid chars', () => {
    rg.policies[0].metadata.name = 'special_not_allowed_$$';
    expect.assertions(1);

    try {
      validateResourceGroupOrThrow(rg);
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });
});

describe('validateMetadataNameConflicts', () => {
  let rg: ResourceGroup;

  beforeEach(() => {
    rg = {
      schemas: [
        { metadata: { namespace: 'some-ns', name: 'schema-name-1' }, schema: 'schema' },
        { metadata: { namespace: 'some-ns', name: 'schema-name-2' }, schema: 'schema' },
      ],
      upstreams: [],
      upstreamClientCredentials: [],
      policies: [
        { metadata: { namespace: 'some-ns', name: 'policy-1' }, type: PolicyType.opa, code: 'code' },
        { metadata: { namespace: 'some-ns', name: 'policy-2' }, type: PolicyType.opa, code: 'code' },
      ],
      remoteSchemas: [],
    };
  });

  it("doesn't throw when multiple resources have the same namespace name", () => {
    expect(() => validateResourceGroupOrThrow(rg)).not.toThrow();
  });

  it('throws when two namespace names are similar, differing only between underscores and dashes', () => {
    rg.policies[1].metadata.namespace = 'some_ns';
    expect.assertions(1);

    try {
      validateResourceGroupOrThrow(rg);
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });

  it('throws when two resource names from the same namespace and resource type are similar, differing only between underscores and dashes', () => {
    rg.policies[1].metadata.name = 'policy_1';
    expect.assertions(1);

    try {
      validateResourceGroupOrThrow(rg);
    } catch (err) {
      expect(err).toMatchSnapshot();
    }
  });

  it("doesn't throw when two resource names from the same namespace and different resource types are similar, differing only between underscores and dashes", () => {
    rg.schemas[1].metadata.name = 'policy_1';
    expect(() => validateResourceGroupOrThrow(rg)).not.toThrow();
  });

  it("doesn't throw when two resource names from the same resource type and different namespaces are similar, differing only between underscores and dashes", () => {
    rg.policies[1].metadata = { namespace: 'other-ns', name: 'policy_1' };
    expect(() => validateResourceGroupOrThrow(rg)).not.toThrow();
  });
});
