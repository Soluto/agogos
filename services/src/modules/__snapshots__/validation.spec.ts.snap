// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateMetadataCharacters - Validation that namespaces and names contain only valid gql chars and dashes throws when namespace names contain invalid chars 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Invalid characters found in namespace name special_not_allowed_@@ of resource type schemas, allowed characters are /^[A-Z_a-z-][\\\\w-]*$/",
      "exception": {}
    }
  ],
  "data": {}
}
`;

exports[`validateMetadataCharacters - Validation that namespaces and names contain only valid gql chars and dashes throws when resource names contain invalid chars 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Invalid characters found in resource name special_not_allowed_$$ of resource type policies, allowed characters are /^[A-Z_a-z-][\\\\w-]*$/",
      "exception": {}
    }
  ],
  "data": {}
}
`;

exports[`validateMetadataNameConflicts throws when two namespace names are similar, differing only between underscores and dashes 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Namespace name conflict found between some-ns and some_ns, they have to either match exactly or have a difference in non-special characters",
      "exception": {}
    }
  ],
  "data": {}
}
`;

exports[`validateMetadataNameConflicts throws when two resource names from the same namespace and resource type are similar, differing only between underscores and dashes 1`] = `
{
  "errors": [
    {
      "name": "GraphQLError",
      "message": "Resource name conflict found between policy-1 and policy_1 of resource type policies in namespace some-ns, they have to either match exactly or have a difference in non-special characters",
      "exception": {}
    }
  ],
  "data": {}
}
`;
